src/app/core/components/left-nav/left-nav-datasource.ts

```diff
@@ -1,137 +1,62 @@
-interface DataSourceUpdate<T extends NavTreeElement> {
-  type: 'update',
-  elementId?: Id, // element to be selected or updated
-  selection: SelectionOpts,
-  elementUpdate?: (el:T) => T, // if set, function to update the element identified by elementId
-}
-interface DataSourceReplacement<T extends NavTreeElement> {
-  type: 'replace',
-  data: NavTreeData<T>,
-}
-interface DataSourceLoading { type: 'loading' }
-interface DataSourceError { type: 'error', error: any }
-type DataSourceChange<T extends NavTreeElement> = DataSourceUpdate<T>|DataSourceReplacement<T>|DataSourceLoading|DataSourceError;
-function dsDeselect<T extends NavTreeElement>(): DataSourceUpdate<T> {
-  return { type: 'update', selection: SelectionOpts.Deselected };
-}
-function dsUpdateElement<T extends NavTreeElement>(id: Id, select: boolean, update: (el:T) => T): DataSourceUpdate<T> {
-  return { type: 'update', selection: select ? SelectionOpts.Selected : SelectionOpts.UnChanged, elementId: id, elementUpdate: update };
-}
-function dsReplaceWith<T extends NavTreeElement>(data: NavTreeData<T>): DataSourceReplacement<T> {
-  return { type: 'replace', data: data };
-}
-function dsLoading(): DataSourceLoading {
-  return { type: 'loading' };
-}
-function dsError(err: any): DataSourceError {
-  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-  return { type: 'error', error: err };
-}
-function isDSLoading<T extends NavTreeElement>(change: DataSourceChange<T>): change is DataSourceLoading {
-  return change.type === 'loading';
-}
-function isDSError<T extends NavTreeElement>(change: DataSourceChange<T>): change is DataSourceError {
-  return change.type === 'error';
-}
-function isDSUpdate<T extends NavTreeElement>(change: DataSourceChange<T>): change is DataSourceUpdate<T> {
-  return change.type === 'update';
-}
-function isDSReplace<T extends NavTreeElement>(change: DataSourceChange<T>): change is DataSourceReplacement<T> {
-  return change.type === 'replace';
-}
-
-
 export abstract class LeftNavDataSource<ContentT extends RoutedContentInfo, MenuT extends NavTreeElement> {
 
   private initialized = false;
-  private contentChanges = new Subject<[content: ContentT|undefined, state: FetchState<NavTreeData<MenuT>>]>();
+  private contentChanges = new ReplaySubject<ContentT|undefined>(1);
   private retryTrigger = new Subject<void>();
-  state: FetchState<NavTreeData<MenuT>> = fetchingState();
-
-  constructor() {
-
-    this.contentChanges.pipe(
-
-      repeatLatestWhen(this.retryTrigger),
-
-      mergeMap(([ contentInfo, prevState ]):ObservableInput<DataSourceChange<MenuT>> => {
-
-        if (prevState.isReady) {
-          // CASE: the current content is not an item and the menu has already items displayed
-          if (!contentInfo) {
-            if (prevState.data.selectedElementId !== undefined) return of(dsDeselect<MenuT>());
-            return EMPTY; // no change
-          }
+  state$ = this.contentChanges.pipe(
```

Comment:
Here there is a accumulation of nested and sibling `if`s, I think/hope there is a way to declare each scenario in a private observable, and `state$` would be a `merge(…)` of them.
It could also be achieved by calling private methods in the `switchScan` callback, the point is to reduce the cognitive load.
What do you think ?

Reply:
In the current version, the content of the `switchScan` block looks quite good to me to understand the overall logic (14 lines of code, the rest are comments): you see that based on the previous state and the new "content", there is a decision to be made on what to be displayed. And the action based on these case are quite straightforward so reading them directly (details are hidden in external functions already) looks clear to me.
I don't see exactly what is your idea with your merge of observable,  but doing purer Rx here (as often) would not make the logic more readable IMO. I think it would be the opposite even. But I might be wrong :-)